Make the SDK friendly to locally link and develop on

MadLittleMods opened the pull request:
Make the `hydrogen-view-sdk` friendly to locally link and develop on. Can now simply `yarn watch:sdk` to watch and build on changes and use in CommonJS and ESM environments.

Also adds `yarn test:sdk` to test whether the SDK works in an ESM Vite build and when using `require` in Node.js (CommonJS).

Fix https://github.com/vector-im/hydrogen-web/issues/686
Fix https://github.com/vector-im/hydrogen-web/issues/682
Fix https://github.com/vector-im/hydrogen-web/issues/722

Instead of deleting the whole `target/` directory, leave it alone so the symlink driving the `npm link`/`yarn link` stays in tact.

Leave Vite builds in their build directories (`/lib-build`/`/asset-build`) so you can `vite build --watch` to build on local changes and still have a consisent place to reference in the `package.json` `exports`. Previously, everything relied on `build.sh` which does a bunch of moving and renaming and made it hard to quickly rebuild on changes.

Add back support for CommonJS (adding the `package.json` `exports`).

The last piece is making sure the `?url` imports (ex. `import workerPath from 'hydrogen-view-sdk/main.js?url';`) still work. ~~It looks like this may have just been solved via https://github.com/vitejs/vite/issues/6725 -> https://github.com/vitejs/vite/pull/7073 (literally 2 days ago) and we just need to wait for the next Vite release~~ Was solved by https://github.com/vitejs/vite/pull/7098 and is available in the latest Vite releases 🎉

Also updating `vite` because it includes fixes for https://github.com/vector-im/hydrogen-web/issues/722 which were added in https://github.com/vitejs/vite/issues/7822 ->https://github.com/vitejs/vite/pull/7827


## Dev notes

 - `package.json` and `exports` docs: https://nodejs.org/docs/latest-v17.x/api/packages.html
 - https://vitejs.dev/config/#build-assetsdir



#### Use the latest Vite from GitHub source (before release)

How can I try out the latest Vite without waiting for a release?

See https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#repo-setup to build locally and `npm link` it

---

How to yarn install a package from a GitHub Monorepo? This doesn't work for Vite as it requires more build stuff.

See

 - https://github.com/yarnpkg/yarn/issues/4725#issuecomment-761528734
 - https://github.com/yarnpkg/yarn/issues/1570

```
yarn add https://gitpkg.now.sh/vitejs/vite/packages/vite?main --dev
```

MadLittleMods committed:
Make the SDK friendly to locally link and develop on

Fix https://github.com/vector-im/hydrogen-web/issues/686
Fix https://github.com/vector-im/hydrogen-web/issues/682

Instead of deleting the whole `target/` directory, leave it alone so the symlink
driving the `npm link`/`yarn link` stays in tact.

Leave Vite builds in their build directories (`/lib-build`/`/asset-build`)
so you can `vite build --watch` to build on local changes and still have a
consisent place to reference in the `package.json` `exports`. Previously,
everything relied on `build.sh` which does a bunch of moving and renaming
and made it hard to rebuild on changes.

Add back support for CommonJS (adding the `package.json` `exports`).

The last piece is making sure the `?url` imports (`import workerPath from 'hydrogen-view-sdk/main.js?url';`)
work still. It looks like this may have just been solved via
https://github.com/vitejs/vite/issues/6725 -> https://github.com/vitejs/vite/pull/7073
(literally 2 days ago) and we just need to wait for the next Vite release 🎉

MadLittleMods commented on scripts/sdk/build.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-rm -rf target
+rm -rf target/*
```
Only remove the directory contents instead of the whole directory to maintain the `npm link`/`yarn link` symlink

MadLittleMods replied:
Added as a comment to the script

MadLittleMods commented on scripts/sdk/build.sh:
```diff
@@ -10,13 +10,7 @@ mkdir target/paths
-mv download-sandbox.*.html ../../download-sandbox.html
-rm *.js *.wasm
-mv ./* ../../
+rm !(main).js *.wasm
```
We want to remove all `*.js` files except for `main.js`

bwindels replied:
can you add a comment please as this syntax is not super common

MadLittleMods commented on scripts/sdk/build.sh:
```diff
@@ -10,13 +10,7 @@ mkdir target/paths
 pushd asset-build/assets
-mv main.*.js ../../main.js
-mv index.*.css ../../style.css
-mv download-sandbox.*.html ../../download-sandbox.html
```
Instead of needing to move these 3 files and rename them without the hash here, we handle removing the hash in `vite.sdk-assets-config.js` (`rollupOptions.output.assetFileNames`) and reference them directly in place in the `package.json` `exports`

MadLittleMods commented on scripts/sdk/base-manifest.json:
```diff
@@ -2,6 +2,17 @@
+    "exports": {
+        ".": {
+            "import": "./lib-build/hydrogen.es.js",
+            "require": "./lib-build/hydrogen.cjs.js"
```
Referencing `hydrogen.es.js` directly in the build, `./lib-build/hydrogen.es.js`, to make it compatible with watching for an local changes and rebuild: `yarn run vite build -c vite.sdk-lib-config.js --watch`

MadLittleMods commented on scripts/sdk/base-manifest.json:
```diff
@@ -2,6 +2,17 @@
+        "./style.css": "./asset-build/assets/index.css",
+        "./main.js": "./asset-build/assets/download-sandbox.html",
+        "./download-sandbox.html": "./asset-build/assets/download-sandbox.html",
+        "./assets/*": "./asset-build/assets/*"
```
Fix https://github.com/vector-im/hydrogen-web/issues/682

Make the assets path available to reference and serve separately.

```js
app.use(express.static(path.dirname(require.resolve('hydrogen-view-sdk/assets/main.js'))));
```

MadLittleMods commented on scripts/sdk/base-manifest.json:
```diff
@@ -2,6 +2,17 @@
+            "import": "./lib-build/hydrogen.es.js",
+            "require": "./lib-build/hydrogen.cjs.js"
+        },
+        "./paths/vite": "./paths/vite.js",
```
This export isn't used but it's part of the `./scripts/sdk/transform-paths.js ./src/platform/web/sdk/paths/vite.js ./target/paths/vite.js` stuff in `build.sh` that's currently commented out.

MadLittleMods commented on vite.sdk-assets-config.js:
```diff
@@ -2,10 +2,29 @@ const path = require("path");
+                  // files in our `package.json` `exports`
+                  if(pathsToExport.includes(path.basename(chunkInfo.name))) {
+                    return "assets/[name].[ext]";
+                  }
```
Instead of renaming these files in `build.sh`, we can just give them the proper name as part of the Vite build.

MadLittleMods commented on vite.sdk-assets-config.js:
```diff
@@ -2,10 +2,29 @@ const path = require("path");
+                  }
+
+                return "assets/[name]-[hash][extname]";
+              }
```
For reference, docs for the `rollupOptions.output.assetFileNames` option, https://rollupjs.org/guide/en/#outputassetfilenames

MadLittleMods committed:
Fix typos pointing to wrong files

MadLittleMods committed:
Add a placeholder for upgrading vite to comment on

MadLittleMods commented on package.json:
```diff
@@ -44,7 +44,7 @@
     "text-encoding": "^0.7.0",
     "typescript": "^4.3.5",
-    "vite": "^2.6.14",
+    "vite": "todo: wait for next Vite release",
```
It looked like this may have been solved via https://github.com/vitejs/vite/issues/6725 -> https://github.com/vitejs/vite/pull/7073 (literally 2 days ago) and not in a Vite release yet but it only seems to fix the problem for `*.js?url` or `*.json?url`.

I played around with that fix and it still doesn't fix the situation for `*.html?url` or non-ESM `.js?url` imports. I wrote some tests, see https://github.com/vitejs/vite/pull/7097

MadLittleMods replied:
Another Vite community member created a fix after seeing my tests, https://github.com/vitejs/vite/pull/7098, which is now merged and in the latest Vite release. Need to pull it in and see if it works for sure 🙂

MadLittleMods replied:
Seems to work ✅

MadLittleMods committed:
Merge branch 'master' into madlittlemods/686-682-local-friendly-development-and-commonjs

Conflicts:
	package.json
	scripts/sdk/base-manifest.json

MadLittleMods committed:
Update Vite which includes fixes to importing `*.js?url` with `exports`

Update to Vite which includes https://github.com/vitejs/vite/pull/7098

MadLittleMods committed:
Avoid ERR_REQUIRE_ESM errors when requiring SDK

MadLittleMods commented on package.json:
```diff
@@ -13,7 +13,8 @@
     "build": "vite build",
-    "build:sdk": "./scripts/sdk/build.sh"
+    "build:sdk": "./scripts/sdk/build.sh",
+    "watch:sdk": "./scripts/sdk/build.sh && yarn run vite build -c vite.sdk-lib-config.js --watch"
```
Any preference on the script name for this? `yarn watch:sdk`

MadLittleMods commented on package.json:
```diff
@@ -46,7 +47,7 @@
     "text-encoding": "^0.7.0",
     "typescript": "^4.3.5",
-    "vite": "^2.6.14",
+    "vite": "^2.9.1",
```
Updating Vite to include fix from https://github.com/vitejs/vite/pull/7098 which allows `import workerPath from 'hydrogen-view-sdk/main.js?url';` to still work even though we have `package.json` `exports` defined.

MadLittleMods replied:
Re-commented on the diff since this one is outdated, https://github.com/vector-im/hydrogen-web/pull/693#discussion_r866272963

MadLittleMods committed:
Add way to test whether SDK works in ESM and CommonJS

MadLittleMods commented on package.json:
```diff
@@ -11,9 +11,11 @@
     "lint-ci": "eslint src/",
     "test": "impunity --entry-point src/platform/web/main.js src/platform/web/Platform.js --force-esm-dirs lib/ src/ --root-dir src/",
     "test:postcss": "impunity --entry-point  scripts/postcss/test.js ",
+    "test:sdk": "cd ./scripts/sdk/test/ && yarn --no-lockfile && node test-sdk-in-esm-vite-build-env.js && node test-sdk-in-commonjs-env.js",
```
This one is a bit clunky but it makes it easy to test whether `hydrogen-view-sdk` actually works in an ESM Vite build and when using `require` in Node.js (CommonJS).

All of the "testing" code was added in https://github.com/vector-im/hydrogen-web/pull/693/commits/2401b7f453a417dc87635e43aa2339ed2517600c if you want to review it on its own. Feel free to prompt to remove or different way to test.

Thought the testing was slightly needed as testing manually in each environment is cumbersome. And we really don't know whether the SDK actually works before pushing it out. All of the usage code is just documented in `SDK.md`

MadLittleMods committed:
Merge branch 'master' into madlittlemods/686-682-local-friendly-development-and-commonjs

Conflicts:
	package.json
	scripts/sdk/base-manifest.json
	scripts/sdk/build.sh

MadLittleMods commented on scripts/sdk/build.sh:
```diff
@@ -10,17 +10,7 @@ mkdir target/paths
-mv download-sandbox.*.html ../../download-sandbox.html
-rm *.js *.wasm
-mv ./* ../../
+rm !(main).js *.wasm
```
This needs `shopt -s extglob` to enable extended globs.

Seems to work fine with git bash on Windows by default but had issues on macOS

```
./scripts/sdk/build.sh: line 13: syntax error near unexpected token `('
./scripts/sdk/build.sh: line 13: `rm !(main).js *.wasm'
```

MadLittleMods committed:
Enable extended globs for removing all but some filename !(filename)

See https://github.com/vector-im/hydrogen-web/pull/693#discussion_r853534719

MadLittleMods commented on scripts/postcss/css-url-processor.js:
```diff
@@ -39,7 +39,7 @@ function colorsFromURL(url, colorMap) {
     const value = decl.value;
     const parsed = valueParser(value);
-    parsed.walk(async node => {
+    parsed.walk(node => {
```
This change will disappear once https://github.com/vector-im/hydrogen-web/pull/721 merges

MadLittleMods commented on scripts/postcss/svg-colorizer.js:
```diff
@@ -37,7 +37,7 @@ module.exports.buildColorizedSVG = function (svgLocation, primaryColor, secondar
         throw new Error("svg-colorizer made no color replacements! The input svg should only contain colors #ff00ff (primary, case-sensitive) and #00ffff (secondary, case-sensitive).");
     }
-    const fileName = svgLocation.match(/.+\/(.+\.svg)/)[1];
+    const fileName = svgLocation.match(/.+[/\\](.+\.svg)/)[1];
```
This change will disappear once https://github.com/vector-im/hydrogen-web/pull/721 merges

MadLittleMods committed:
Fix tests after theme updates

MadLittleMods commented on package.json:
```diff
@@ -11,9 +11,11 @@
     "lint-ci": "eslint src/",
     "test": "impunity --entry-point src/platform/web/main.js src/platform/web/Platform.js --force-esm-dirs lib/ src/ --root-dir src/",
     "test:postcss": "impunity --entry-point  scripts/postcss/tests/css-compile-variables.test.js scripts/postcss/tests/css-url-to-variables.test.js",
+    "test:sdk": "yarn build:sdk && cd ./scripts/sdk/test/ && yarn --no-lockfile && node test-sdk-in-esm-vite-build-env.js && node test-sdk-in-commonjs-env.js",
```
This one is a bit clunky but it makes it easy to test whether `hydrogen-view-sdk` actually works in an ESM Vite build and when using `require` in Node.js (CommonJS).

All of the "testing" code was added in https://github.com/vector-im/hydrogen-web/pull/693/commits/2401b7f453a417dc87635e43aa2339ed2517600c if you want to review it on its own. Feel free to prompt to remove or different way to test.

Thought the testing was slightly needed as testing manually in each environment is cumbersome. And we really don't know whether the SDK actually works before pushing it out. All of the usage code is just documented in `SDK.md`

MadLittleMods commented on package.json:
```diff
@@ -47,7 +49,7 @@
     "text-encoding": "^0.7.0",
     "typescript": "^4.3.5",
-    "vite": "^2.6.14",
+    "vite": "^2.9.1",
```
Now with the [recent theming updates](https://github.com/vector-im/hydrogen-web/pull/709), it added a custom PostCSS plugin which doesn't play well with `vite@2.9.0`+, see https://github.com/vector-im/hydrogen-web/issues/722

We will probably need to resolve that situation before merging this PR.

MadLittleMods replied:
This was fixed in Vite via https://github.com/vitejs/vite/issues/7822 ->https://github.com/vitejs/vite/pull/7827 and is part of `vite@2.9.8` 🎉

MadLittleMods replied:
Updated in PR

bwindels commented on scripts/sdk/base-manifest.json:
```diff
@@ -1,15 +1,17 @@
         "./paths/vite": "./paths/vite.js",
-        "./style.css": "./asset-build/assets/index.css",
+        "./style.css": "./asset-build/assets/theme-element-light.css",
+        "./theme-element-light.css": "./asset-build/assets/theme-element-light.css",
```
Could we avoid this if we just say in the SDK docs that the path to the css file is `hydrogen-view-sdk/assets/theme-element-light.css` rather than `hydrogen-view-sdk/theme-element-light.css`. Will be easier towards the future when adding more assets. Probably best to keep style.css for now for backwards compat though.

MadLittleMods replied:
> Could we avoid this if we just say in the SDK docs that the path to the css file is `hydrogen-view-sdk/assets/theme-element-light.css` rather than `hydrogen-view-sdk/theme-element-light.css`. Will be easier towards the future when adding more assets. 

Yes, we can! Should we do that in this PR or in a follow-up?

MadLittleMods replied:
The 👍 reaction is ambiguous on what option to do so I'm going to opt to fix this in a fast follow-up PR.

bwindels replied:
wfm

MadLittleMods replied:
Addressed in https://github.com/vector-im/hydrogen-web/pull/746 ⏩

bwindels reviewed (commented):
This is fantastic work, thank you so much for all your effort! And sorry for totally missing this PR before.
The changes look great, I just need to run the build on this branch to verify that it all still works, but I'll do that once the blockers are resolved, AIUI:
 - #721 needs to merge
 - https://github.com/vector-im/hydrogen-web/issues/722 is should be fixed

I've asked @MidhunSureshR to have a look.

MadLittleMods committed:
Merge branch 'master' into madlittlemods/686-682-local-friendly-development-and-commonjs

Conflicts:
	scripts/sdk/build.sh

MadLittleMods committed:
Explain what is being deleted by the strange syntax

See https://github.com/vector-im/hydrogen-web/pull/693#discussion_r815284713

MadLittleMods committed:
Remove extra space

MadLittleMods committed:
Update Vite to avoid flakey errors in our PostCSS plugins

Fix https://github.com/vector-im/hydrogen-web/issues/722

Updating Vite to includes fixes from
https://github.com/vitejs/vite/issues/7822 -> https://github.com/vitejs/vite/pull/7827

MadLittleMods commented on package.json:
```diff
@@ -47,7 +49,7 @@
     "text-encoding": "^0.7.0",
     "typescript": "^4.3.5",
-    "vite": "^2.6.14",
+    "vite": "^2.9.8",
```
Updating Vite to include fix from https://github.com/vitejs/vite/pull/7098 which allows `import workerPath from 'hydrogen-view-sdk/main.js?url';` to still work even though we have `package.json` `exports` defined (`vite@2.9.1`).

And to includes fixes for https://github.com/vector-im/hydrogen-web/issues/722 which were added in https://github.com/vitejs/vite/issues/7822 ->https://github.com/vitejs/vite/pull/7827 (`vite@2.9.8`)

MadLittleMods committed:
Merge branch 'master' into madlittlemods/686-682-local-friendly-development-and-commonjs

MadLittleMods committed:
Add some comments

MadLittleMods committed:
Merge branch 'master' into madlittlemods/686-682-local-friendly-development-and-commonjs

Conflicts:
	scripts/sdk/base-manifest.json

bwindels reviewed (commented):
Thanks, changes look great, but when running `yarn build:sdk` or `yarn watch:sdk`, I get this error:
```
$ yarn watch:sdk
yarn run v1.22.17
$ ./scripts/sdk/build.sh && yarn run vite build -c vite.sdk-lib-config.js --watch
$ /home/bwindels/dev/hydrogen-web/node_modules/.bin/vite build -c vite.sdk-assets-config.js
Generated an empty chunk: "vendor"
Generated an empty chunk: "index"
$ /home/bwindels/dev/hydrogen-web/node_modules/.bin/vite build -c vite.sdk-lib-config.js
$ /home/bwindels/dev/hydrogen-web/node_modules/.bin/tsc -p tsconfig-declaration.json
~/dev/hydrogen-web/target ~/dev/hydrogen-web
~/dev/hydrogen-web/target/asset-build/assets ~/dev/hydrogen-web/target ~/dev/hydrogen-web
~/dev/hydrogen-web/target ~/dev/hydrogen-web
rm: cannot remove 'index.html': No such file or directory

MadLittleMods committed:
Clean up index.html in the right spot

MadLittleMods commented on scripts/sdk/build.sh:
```diff
@@ -12,19 +16,10 @@ mkdir target/paths
-mv download-sandbox.*.html ../../download-sandbox.html
-rm *.js *.wasm
-mv ./* ../../
+pushd target/asset-build
+rm index.html
```
> [...] when running `yarn build:sdk` or `yarn watch:sdk`, I get this error:
>
> ```
> [...]
> rm: cannot remove 'index.html': No such file or directory
> ```
>
> *-- https://github.com/vector-im/hydrogen-web/pull/693#pullrequestreview-973791417*

Fixed this up by removing `index.html` in the right spot (`target/asset-build/index.html`).

Sorry didn't pick up on this earlier. I think it's because when running on Windows the git bash window which pops up when you `yarn build:sdk` to run the shell code closes so fast and masks errors for me. I've tested this on macOS and everything is working now 👍

bwindels reviewed (approved):
Works now indeed, thanks!

bwindels committed:
Merge branch 'master' into madlittlemods/686-682-local-friendly-development-and-commonjs
